RFC(Accessibility): Converged accessibility (#16988)
microsoft/fluentui: rfcs/accessibility/accessibility.md

# RFC: Accessibility

---

Contributors:
@jurokapsiar

## Summary

Fluent UI components are designed to support various aspects of accessibility, so that they can be used with different input methods (mouse, touch, keyboard, screen readers) as well as fulfill different rendering and layout requirements (theming, zoom, contrast).

## Background

[Web Content Accessibility Guidelines (WCAG)](https://www.w3.org/TR/WCAG21/) and [WAI-ARIA Authoring Practices](https://www.w3.org/TR/wai-aria-practices-1.2/) are community driven standard for accessible web pages and applications. They provide basic framework allowing consistent usage of web pages and applications with or without assistive technologies.

## Problem statement

> Fluent UI components aim to fully respect both WCAG and ARIA and the Fluent UI team is working with the community to improve them. From the technical point of view, thee main challenges were identified:

Reviewer: just typo

> 1. Accessibility is usually not intuitive as it is not immediately obvious. It is hard to achieve high quality user experience and easy to regress it.

Reviewer 2: I completely get what this point is trying to capture, though I think the current wording risks implying that the only perspective we're speaking to is an abled one. That is to say, accessibility is very obvious and pretty intuitive for someone who is disabled. Maybe it could be reworded a bit to still capture the challenge without the implication?

One option could be something like this:
"Accessibility is a complex field that is often not taught and left behind in planning, design, development, and testing. It can be hard to achieve high quality user experiences and easy to regress them."

> 2. Each screen reader and browser works differently, often making it hard or even impossible to achieve high quality user experience for all combinations. Browsers and screen readers release new versions that change the experience

Reviewer: `that change` not sure what would be the best but I would rather express this "can" happen, or it is high risk that new releases can bring regression....

Reviewer 3: I agree, maybe something in the lines of

`One of the biggest reasons behind this is that browsers and screen readers are often updated, with each update having the potential to change the user experience from one version to the next.`

> 3. Not all UI/UX designs can be sufficiently represented with ARIA roles and patterns

Reviewer 3: nit:
```suggestion
3. Not all UI/UX designs can be sufficiently represented with the existing ARIA roles and patterns.
```

## Detailed Design or Proposal

> To address these challenges, following principles will be followed:

Reviewer 3: nit:
```suggestion
To address these challenges, the following principles will be followed:
```

> - Challenge 1: Sufficient test coverage with understandable and documented tests
> - Challenge 2: Components will be created based on well tested prototypes or existing components. We will report and track issues to the screen readers vendors
> - Challenge 3: We will work with the community to improve standards. We will maintain set of best practices and validations that help consumers design accessible user interfaces

Reviewer 3: nit:
```suggestion
- Challenge 1: Sufficient test coverage will be provided, with understandable and documented tests.
- Challenge 2: Components will be created based on well tested prototypes or existing components. We will report and track issues to the screen readers' vendors.
- Challenge 3: We will work with the community to improve standards. We will maintain a set of best practices and validations to help consumers design accessible user interfaces.
```

> ### Target accessibility areas

Reviewer: do we want to add to this section as well "screen reader navigation" ? I mean that component should be accessible when user navigate with Virtual Cursor/Browse mode/Scan mode/VoiceOver keys

- DOM structure that provides semantical value either by using correct element types or roles
- ARIA attributes that are valid for elements/roles used and provide correct state information about the component or element

> - Keyboard navigation (navigation by tabbing, arrow keys, pagination or letter keys, click and right click (Enter/Space, Shift+F10), and close (Escape))

Reviewer: Maybe I would just add something like. 
Keyboard navigation is applied based on component.  (In order to avoid confusion that in every component has the same keyboard navigation expectation)

> - Focus handling when the component is able to move the focus in a predictable way - mostly when opening menus, popups or dialogs or dismissing them using Esc. Focus trap for Dialog and popups

Reviewer 4: Should we mention anyhow autofocus behavior?

> - Sufficient contrast level

Reviewer: maybe I would rather add  `color contrast`
at least Accessibility insight call it in this way
![image](https://user-images.githubusercontent.com/22620150/107943710-6033a000-6f8d-11eb-9516-9786c4a626be.png)

- Light, Dark and High Contrast themes
- Displaying focus indicator when keyboard is used to interact with them

> - Zoom up to 400%

Reviewer 3: I would maybe phrase this one in the lines of something like:

```suggestion
- Tested against visual inconsistencies/bugs on a zoom up to 400%
```

> - Zoom up to 400%

Reviewer 5: Could we explicitly mention that only browser zoom is supported, not font-size-based scaling?

> 

Reviewer 4: Do we target anything related to mobile devices? Touch interactions are a good example of this, they are different than on desktops.

Reviewer 2: Touch is also a good proxy for a number of assistive tech interactions, e.g. switch + mouse grid, eye gaze, etc.

Fluent UI components will use [ability-helpers](https://github.com/microsoft/ability-helpers) for focus handling functionality, so that they can be easily integrated with application-level ability-helpers functionality such as delooser and cross-iframe focusing.

### Out of scope

> Internationalization, globalization, keyboard shortcuts and language detection are deliberately not part of Fluent UI and should be handled by the hosting application.

Reviewer 4: I just would to mention that React Spectrum handles it, for example: https://react-spectrum.adobe.com/react-spectrum/Provider.html#locales

May be at some point we will reconsider this statement :)

Author: maybe - currently we had props for either texts or formatting callbacks. We can consider allowing consumers to provide a dictionary, but I would like to avoid owning any particular text translations or formatting decisions as they might be dependent on different things like OS environment.

Focus handling (except of the points mentioned above) on an application level needs to be handled by the application, preferably using [ability-helpers](https://github.com/microsoft/ability-helpers).

### Component development process

1. Fill in Accessibility section of the [component specification](https://github.com/microsoft/fluentui/blob/master/scripts/create-package/plop-templates-react/Spec.md.hbs)
2. Create prototype or base functionality of the component or reference an existing Fluent UI React or Fluent UI React Northstar component as model example
3. Perform an initial test on prototype, model component or when converged component functionality is available (all supported screen reader/browser combinations)
4. Incorporate test results in the component implementation
5. Create [component behavior tests](https://github.com/microsoft/fluentui/tree/master/packages/a11y-testing/src/definitions)

> 6. Document best practices

Reviewer 4: ```suggestion
6. Document best practices (includes Do's and Don'ts)
```

?

7. Document known issues, including links if they are tracked by browsers or screen readers
8. Create validation rules for component - help to identify accessibility issues early in the development phase (or even design with UI Builder)

> 9. Schedule trusted user test

Reviewer 6: Love this! We should add this to the "integration/validation" phase in the component template

- Challenge 3: We will work with the community to improve standards. We will maintain a set of best practices and validations that help consumers design accessible user interfaces.

### Target accessibility areas

- DOM structure that provides semantical value either by using correct element types or roles
- ARIA attributes that are valid for elements/roles used and provide correct state information about the component or element
- Keyboard navigation (navigation by tabbing, arrow keys, pagination or letter keys, click and right click (Enter/Space, Shift+F10), and close (Escape)) applied based on the component
- Screen reader navigation (Virtual Cursor/Browse mode/Scan mode/VoiceOver keys)
- Touch interaction
- Focus handling when the component is able to move the focus in a predictable way - mostly when opening menus, popups or dialogs (autofocus) or dismissing them using Esc. Focus trap for Dialog and popups
- Sufficient color contrast
- Light, Dark and High Contrast themes
- Displaying focus indicator when keyboard is used to interact with them
- Tested against visual inconsistencies/bugs on a zoom up 400%

> Fluent UI components will use [ability-helpers](https://github.com/microsoft/ability-helpers) for focus handling functionality, so that they can be easily integrated with application-level ability-helpers functionality such as delooser and cross-iframe focusing.

Reviewer 6: ```suggestion
Fluent UI components will use [tabster](https://github.com/microsoft/tabster) for focus handling functionality, so that they can be easily integrated with application-level tabster functionality such as delooser and cross-iframe focusing.
```
Might just want to update for the sake of getting the new name out there

Reviewer 6: not 100% sure if the 3rd instance of ability-helpers should be replaced with tabster, i'll leave it up to you
